common/models/transducer/multi_stager.py

```diff
@@ -0,0 +1,284 @@
+
+from returnn.tf.util.data import Data, DimensionTag
+from returnn.config import get_global_config
+from ...training.switchout import switchout_target
+from ...datasets.interface import TargetConfig
+from .topology import Topology, rna_topology, rnnt_topology
+
+import tensorflow as tf
+import sys
+import os
+
+
+
+class Context:
+  def __init__(self, task: str, target: TargetConfig, model: str,
+               name: str, alignment_topology: Topology = rnnt_topology):
+    self.task = task
+    self.train = (task == "train")
+    self.search = (task != "train")
+    self.target = target
+    self.align_dir = os.path.dirname(model)
+    self.name = name
+    self.num_labels_nb = target.get_num_classes()
+    self.num_labels_wb = self.num_labels_nb + 1
+    self.blank_idx = self.num_labels_nb
+    self.alignment_topology = alignment_topology
+
+
+def make_align(net: Dict[str, Any],
+               epoch: int,  # maybe required
+               extend: bool,
+               ctx: Context,
+               output: str = "output",
+               output_log_prob: str = "output_log_prob_wb",
+               encoder: str = "encoder",
+               target: TargetConfig = None):
+  """
+  Here we assume that the decoder is a recurent network(with unit) called `output`.
+  In the "unit" `output_log_prob` should define the the log distribution over the whole vocab inkl blank.
+  Otherwise "base:{encoder}" which represent the output of the encoder should be provided.
+  This function extends the "unit" of the decoder with logic to create and dump fullsum alginment in .hdf files.
+  Requires:
+    output/unit
+      output_log_prob: log distribution over the whole vocab inkl blank
+      f"base:data:{target}": targets of the sequence
+      base:encoder: output of the encoder
+    rna_fullsum_alignment: function that performs the alignment and returns for e.g [BxT] for rna alignm.
+    extend: if True the already existing .hdf alignments are extended
+    ctx: holds information such as the label topology, target and path to be used for .hdf files
+
+  Durign this step 1 subepoch passes.
+  """
+  align_dir = ctx.align_dir
+  name = ctx.name
+  if not target:
+    target = TargetConfig.global_from_config()
+  subnet = net[output]["unit"]
+  subnet[output_log_prob]["is_output_layer"] = True
+  subnet["fullsum_alignment"] = {
+    "class": "eval",
+    "from": [output_log_prob, f"base:data:{ctx.target.key}", f"base:{encoder}"],
+    "eval": ctx.alignment_topology.alignment,
+    "out_type": ctx.alignment_topology.alignment_out_type,
+    "is_output_layer": True
+  }
+
+  subnet["_align_dump"] = {
+    "class": "hdf_dump",
+    "from": "fullsum_alignment",
+    "is_output_layer": True,
+    "dump_per_run": True,
+    "extend_existing_file": extend,  # TODO: extend only the first time
+    # dataset_name: comes from **opts of the lambda in filename
+    "filename":
+      (lambda **opts: "{align_dir}/align.{name}_{dataset_name}.hdf".format(align_dir=align_dir,
```

Comment:
Do not use `str.format`. Use f-strings.

Reply:
I don't think it is easy to do in this case. How would you do it? the `dataset_name` comes from `**opts`.